### DISCUSSION TITLE
Outbound traffic from Pods does not honor ip_local_port_range

### DISCUSSION THREAD

**Discussion by torbenaa** [k3s-io/k3s#6720] [Q&A]
Running v1.25.3-k3s1 on ubuntu 22.04.01 LTS in Hetzner Robot.

Unfortunatly, Hetzner does not provide a statefull firewall for their robot servers - so I have to protect my K3S server with basic L3 switch ACL's to prevent access to the exposed ports on the server. I have been advised NOT to mangle with the iptables chains to protect inbound traffic.

Hetzner only filter incoming packets. So I need to allow replies from the servers I am accessing on the Internet.To implement this, I have verified the Ephemeral  port range ( from /proc/sys/net/ipv4/ip_local_port_range) and created an ACL that allows traffic to these ports.

Everything works fine for traffic origination from the linux host itself.

However - traffic originating from the Pods running on the node is being source NAT'ed to ports outside this range ( 32768-65535 ). 

Should K3S honor those settings when doing iptables MASQ ?

Thank you very much !

Best Regards
Torben

**Comment by brandond** [comment]
Traffic coming from pods towards outside networks would be subject to SNAT since the actual pod IPs are not routable, so no you will not see the original source port when it leaves the host.

**Comment by torbenaa** [comment]
Hi Brad,

Thanks for the prompt reply. I understand that source NAT (masquerade) is needed for packets to leave the host with a routable ip-adresse.
But I would expect that the iptables implementation would honor the defined source port ranges on the host.
Apparently, it is not.

Using traditional iptables rules, one can define the port range using "--to-ports" like this : 

iptables -t nat -A POSTROUTING -p TCP -j MASQUERADE --to-ports 32768-65535.

How can I define this range in K3S ?

Best Regards
Torben

**Reply by brandond** [reply]
>  I would expect that the iptables implementation would honor the defined source port ranges on the host.

That is not true. I'm not super well versed in the ins and outs of the linux kernel's masquerade module, but there is some discussion of this at https://github.com/aws/amazon-vpc-cni-k8s/issues/509:
> When selecting a port for the outgoing connection the kernel `netfliter` module will select the next port avaliable in the `/proc/sys/net/ipv4/ip_local_port_range` range by default. However, due to issues with port conflicts occuring as discussed in #246 the default behaviour has been changed to use the `NF_NAT_RANGE_PROTO_RANDOM` flag by way of the `iptables` `--random` option.
> 
> It turns out that the kernel is hard coded[1] when either the `NF_NAT_RANGE_PROTO_RANDOM` or `NF_NAT_RANGE_PROTO_RANDOM_FULLY` to select a random port in the non-privileged range, i.e. 1024 through 65535.

the `--random/--random-fully` options are used extensively by most projects, so that is the behavior I would expect to see from the various NAT rules generate by Kubernetes components.

Your suggestion to use the `--to-ports` option to restrict this is a good one, but flannel does not have support for that as far as I can tell. Nor do the kubelet or kube-proxy. You could start by asking for this over at https://github.com/flannel-io/flannel?

**Reply by torbenaa** [reply]
Iptables sure is a mess :)

So - it is flannel that is responsible for the iptables entries ?

If so, I will give it a try with the flannel  guys.

Thanks !

**Comment by MTSym** [comment]
@torbenaa 

I do have currently the same problem with k3s on a Hetzner server. Were you able to solve that problem? And if so would you be willing to tell us how you solved it?

From the pods `veth` to `cni` everything is using ports within the ephemeral port range but from `cni` to `enp6s18` it is broken.
```bash
18:11:18.799072 veth07dea84d P   IP (tos 0x0, ttl 64, id 10594, offset 0, flags [DF], proto TCP (6), length 60)
    10.42.2.103.33608 > 116.202.234.38.443: Flags [S], cksum 0x6bb0 (incorrect -> 0x7182), seq 3135374529, win 65500, options [mss 1310,sackOK,TS val 588710903 ecr 0,nop,wscale 7], length 0
18:11:18.799075 cni0  In  IP (tos 0x0, ttl 64, id 10594, offset 0, flags [DF], proto TCP (6), length 60)
    10.42.2.103.33608 > 116.202.234.38.443: Flags [S], cksum 0x6bb0 (incorrect -> 0x7182), seq 3135374529, win 65500, options [mss 1310,sackOK,TS val 588710903 ecr 0,nop,wscale 7], length 0
18:11:18.799085 enp6s18 Out IP (tos 0x0, ttl 63, id 10594, offset 0, flags [DF], proto TCP (6), length 60)
    192.168.1.4.19439 > 116.202.234.38.443: Flags [S], cksum 0x20cc (incorrect -> 0xf3bf), seq 3135374529, win 65500, options [mss 1310,sackOK,TS val 588710903 ecr 0,nop,wscale 7], length 0
```

Thanks in advance for any help on this topic.

**Reply by brandond** [reply]
As mentioned above, this will need to be addressed on the flannel side.
